src/Features/CSharp/Portable/RemoveAsyncModifier/CSharpRemoveAsyncModifierCodeFixProvider.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.ConvertSwitchStatementToExpression;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Formatting;
+using Microsoft.CodeAnalysis.CSharp.MakeMethodSynchronous;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.RemoveAsyncModifier;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using static Microsoft.CodeAnalysis.MakeMethodAsynchronous.AbstractMakeMethodAsynchronousCodeFixProvider;
+
+namespace Microsoft.CodeAnalysis.CSharp.RemoveAsyncModifier
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.MakeMethodSynchronous), Shared]
+    [ExtensionOrder(After = PredefinedCodeFixProviderNames.AddImport)]
+    internal partial class CSharpRemoveAsyncModifierCodeFixProvider : AbstractRemoveAsyncModifierCodeFixProvider
+    {
+        private const string CS1998 = nameof(CS1998); // This async method lacks 'await' operators and will run synchronously.
+
+        [ImportingConstructor]
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
+        public CSharpRemoveAsyncModifierCodeFixProvider()
+        {
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(CS1998);
+
+        protected override bool IsAsyncSupportingFunctionSyntax(SyntaxNode node)
+            => node.IsAsyncSupportingFunctionSyntax();
+
+        protected override bool CanFix(KnownTypes knownTypes, IMethodSymbol methodSymbolOpt)
+        => !methodSymbolOpt.ReturnType.OriginalDefinition.Equals(knownTypes._iAsyncEnumerableOfTTypeOpt) &&
+           !methodSymbolOpt.ReturnType.OriginalDefinition.Equals(knownTypes._iAsyncEnumeratorOfTTypeOpt);
+
+        protected override SyntaxNode ChangeReturnStatements(SyntaxNode node, SyntaxGenerator generator, KnownTypes knownTypes)
+        {
+            var converter = new ConvertReturnsToTaskRewriter(generator, knownTypes._taskType);
+
+            return node switch
+            {
+                MethodDeclarationSyntax method => FixMethodReturns(method, converter),
+                LocalFunctionStatementSyntax localFunction => FixLocalFunctionReturns(localFunction, converter),
+                AnonymousFunctionExpressionSyntax method => FixAnonymousFunctionReturns(method, converter),
+                _ => node,
+            };
```

Comment:
instead of swithcing on the node here, it feels like you could just ask teh rewriter to fix things up and pass it the node.  it would then dispatch the node accordingly.